Let mkfs.ext3 use the image file directly instead of loopback device

wpoely86 opened the pull request:
By avoiding the loopback device, we do not need root priviledges to create an
image. We give mkfs.ext3 an offset and image size.

This is NOT ready: it's only meant to show that it is possible.

@singularityware-admin

wpoely86 committed:
Let mkfs.ext3 use the image file directly instead of loopback

By avoiding the loopback device, we do not need root priviledges to create an
image. We give mkfs.ext3 an offset and image size.

wpoely86 committed:
Use offset parameter of mkfs.ext3 if it is supported

wpoely86 committed:
Drop output of dd in configure

wpoely86 committed:
Travis: show output of failed tests

wpoely86 committed:
Check exit status of mkfs.ext3 in create

wpoely86 committed:
Only use WEXITSTATUS after checking WIFEXITED

According to the man page of waitpid, you can only call WEXITSTATUS
after you have checked WIFEXITED.

wpoely86 committed:
Make sure final mount point exists

wpoely86 committed:
Always enable --debug when running make test

wpoely86 committed:
Add forgotten changes from last commit

wpoely86 committed:
Correct for dropped privileges

wpoely86 committed:
Merge remote-tracking branch 'origin/lib-refactor' into create-nosuid

* origin/lib-refactor:
  Added copy ability to run as non-root
  Added src/copy to gitignore
  Added copy functionality
  Adding debugging when forking
  Adding more tests
  Add create final mount directory when overlayfs is not enabled
  Support multiple allowed container owners (list)
  Return if binary is running suid (enabled?)
  Add ability to limit containers that can run according to owner
  Removed unused privleged mode code (for now...)
  Minor fix to make sure that singularity.conf is always updated
  Add selftest abilities... needs to be expanded upon
  Include MS_NODEV in image mount
  Merge contributing clause back into license as it should be

gmkurtzer commented on configure.ac:
```diff
@@ -274,12 +274,25 @@ AC_ARG_ENABLE([suid],
 AS_IF([test "x$enable_suid" != "xno"], [
     AC_MSG_RESULT([yes])
-    BUILD_SUID="action-suid create-suid copy-suid expand-suid export-suid import-suid mount-suid"
+    BUILD_SUID="action-suid copy-suid expand-suid export-suid import-suid mount-suid"
```
Can you confirm that if the below check for new enough mkfs.ext3 fails, we need to add create-suid back into the BUILD_SUID list.

wpoely86 replied:
That happens on Line 291: it `$BUILD_SUID` is non-empty, we add `create-suid` to the list:
```
AS_IF([test -n "$BUILD_SUID"], [BUILD_SUID="$BUILD_SUID create-suid"], [])
```

wpoely86 replied:
You can check by looking at the build on Travis: it's uses an old version of mkfs.ext3 so it needs SUID rights to work.

gmkurtzer replied:
Ahh, right you are. Sorry for missing that!

gmkurtzer commented on src/create.c:
```diff
@@ -74,24 +78,45 @@ int main(int argc, char **argv) {
+    mkfs_cmd[3] = strjoin("offset=", int2str(strlength(LAUNCH_STRING, 1024)));
+    mkfs_cmd[4] = strdup(image.path);
+    // pass the correct size of the file in KiB
+    mkfs_cmd[5] = int2str((size*1024*1024-strlength(LAUNCH_STRING, 1024))/1024);
```
Is this necessary?

wpoely86 replied:
If you don't pass it, you get (even with `-q`):
```
Warning: offset specified without an explicit file system size.
Creating a file system with 131072 blocks but this might
not be what you want.
```
So I think it's safer to add it.

wpoely86 committed:
Merge remote-tracking branch 'origin/lib-refactor' into create-nosuid2

* origin/lib-refactor: (125 commits)
  Minor changes for #508
  Fix tests to match ordering
  Fix double dots
  Implement read-only bind mounts to close #451
  Fixing the API for the changes I made. ooopsie.
  Add support for %environment (#504)
  Minor environment file updates!
  remove tmpfile
  Don't do recursive chmod here
  Fix permissions to remove files properly
  re-add chmod to make things writable again
  fix import to properly error out
  we should not call add for user import, all calls go to import.py
  removing script that doesn't exist anymore
  updating docker/python functions to not need ADD
  Fix import.exec for new python interface
  Updates for new python interface
  tweaking variable name to SINGULARITY_CONTENTS instead of layerfile
  Updates for newest python API
  adding new python files
  ...

wpoely86 committed:
Fix two errors because of merging upstream lib-refactor

Currently `singularity -d export image.iso > image.tar` will not work
as `-d` makes singularity print to stdout

wpoely86 committed:
Merge remote-tracking branch 'origin/development' into create-nosuid

* origin/development: (67 commits)
  Clean environment when using -C/--containall flags
  Add -e/--cleanenv runtime action option (ref: #445)
  Moved test
  Cleaned up tests and added more docker tests
  Fix shell debug/verbose messages to goto STDERR (ref #481)
  fix: removed tabs
  fix: removed driver if
  added test
  fix syntax
  Testing a wider print format
  Minor fix for CI
  Clean redundant messages of label additions
  Support optional tests
  Make container writable for removal test
  Added config ownership test
  Fix comments and empty lines in the %file sections (and add tests)
  Finishing up the new test code
  Fixes for runscript and environment overriding what is in the sections and python
  Test updates... WIP
  Fix %labels parsing during bootstrap
  ...

wpoely86 committed:
Restore permission
